File: customs-bond-vat-copilot/src/checks/numeric.js
// সংখ্যাগত auto-check ইঞ্জিন — বন্ড অডিটের মূল reconciliation।
//
// মডিউলে থাকে শুধু ইনপুটের ডিক্লারেটিভ স্পেক (serializable → frontend-এ যায়)।
// প্রকৃত হিসাব এখানে, check id দিয়ে keyed। এভাবে ফর্মুলা একটাই জায়গায়, testable,
// আর মডিউল অবজেক্ট JSON-safe থাকে।
//
// প্রতিটি compute ফাংশন parsed values (Number, না দিলে NaN) নেয়, একটি result দেয়:
//   { status, severity, discrepancy, unit, revenueImplication, observation, breakdown }
// status: "ok" (ব্যত্যয় নেই) | "flag" (অসঙ্গতি) | "insufficient" (তথ্য অসম্পূর্ণ)

const round2 = (n) => Math.round((Number(n) + Number.EPSILON) * 100) / 100;
const bdt = (n) => Number(n || 0).toLocaleString("en-BD");
const q = (n) => round2(n).toLocaleString("en-BD"); // পরিমাণ ফরম্যাট

/** required key গুলোর কোনোটি NaN হলে insufficient; সব ইনপুট খালি হলেও insufficient */
function guard(values, spec) {
  const keys = spec.inputs.map((i) => i.key);
  const anyEntered = keys.some((k) => Number.isFinite(values[k]));
  if (!anyEntered) return { status: "insufficient", reason: "কোনো সংখ্যা দেওয়া হয়নি" };
  const missing = spec.inputs
    .filter((i) => !i.optional && !Number.isFinite(values[i.key]))
    .map((i) => i.label);
  if (missing.length) return { status: "insufficient", reason: `প্রয়োজনীয় তথ্য অনুপস্থিত: ${missing.join(", ")}` };
  return null;
}

const ok = (severity, observation, extra = {}) => ({ status: "ok", severity: severity ?? "low", observation, revenueImplication: 0, ...extra });
const flag = (severity, discrepancy, unit, revenueImplication, observation, breakdown) => ({
  status: "flag", severity, discrepancy: round2(discrepancy), unit,
  revenueImplication: round2(revenueImplication), observation, breakdown,
});

// duty ইনপুট optional; না দিলে 0 ধরে শুধু পরিমাণ-ব্যত্যয় দেখায়।
const duty = (v, key) => (Number.isFinite(v[key]) ? v[key] : 0);

// ---------------- per-check computers ----------------

const COMPUTERS = {
  // ১) আমদানি vs অনুমোদিত entitlement/UP
  "num-entitlement"(v) {
    const excess = v.imported - v.approvedEntitlement;
    if (excess <= 0)
      return ok("low", `আমদানি ${q(v.imported)} একক অনুমোদিত entitlement ${q(v.approvedEntitlement)} একক-এর মধ্যে; entitlement breach নেই।`);
    const rev = excess * duty(v, "dutyPerUnit");
    return flag("high", excess, "একক", rev,
      `অনুমোদিত annual entitlement/UP ${q(v.approvedEntitlement)} একক-এর বিপরীতে আমদানি ${q(v.imported)} একক — অতিরিক্ত ${q(excess)} একক (entitlement breach)। সম্ভাব্য রাজস্ব প্রভাব BDT ${bdt(rev)}।`,
      [{ label: "অনুমোদিত entitlement", value: q(v.approvedEntitlement) }, { label: "আমদানি", value: q(v.imported) }, { label: "অতিরিক্ত", value: q(excess) }]);
  },

  // ২ক) Bill of Entry (আমদানি) vs বন্ড রেজিস্টার লিপিবদ্ধ পরিমাণ
  "num-be-register"(v) {
    const diff = v.beQty - v.registerQty;
    const bd = [
      { label: "B/E আমদানি", value: q(v.beQty) },
      { label: "রেজিস্টারে লিপিবদ্ধ", value: q(v.registerQty) },
      { label: "পার্থক্য", value: q(diff) },
    ];
    if (Math.abs(diff) < 1e-6)
      return ok("low", `B/E ${q(v.beQty)} একক = রেজিস্টারে লিপিবদ্ধ ${q(v.registerQty)} একক; মিল আছে।`, { breakdown: bd });
    if (diff > 0) {
      const rev = diff * duty(v, "dutyPerUnit");
      return flag("high", diff, "একক", rev,
        `Bill of Entry অনুযায়ী আমদানি ${q(v.beQty)} একক, কিন্তু বন্ড রেজিস্টারে লিপিবদ্ধ ${q(v.registerQty)} একক — ${q(diff)} একক রেজিস্টারভুক্ত হয়নি (unrecorded raw material; সম্ভাব্য অপসারণ)। সম্ভাব্য রাজস্ব প্রভাব BDT ${bdt(rev)}।`,
        bd);
    }
    return { ...flag("medium", -diff, "একক", 0,
      `বন্ড রেজিস্টারে লিপিবদ্ধ ${q(v.registerQty)} একক B/E আমদানি ${q(v.beQty)} একক-এর চেয়ে ${q(-diff)} একক বেশি — over-recording/রেকর্ড অসঙ্গতি; যাচাই করুন। স্বয়ংক্রিয় রাজস্ব ধরা হয়নি।`,
      bd) };
  },

  // ২) প্রকৃত ব্যবহার vs অনুমোদিত input-output coefficient
  "num-coefficient"(v) {
    const allowed = v.finishedProduced * v.coeffPerUnit;
    const excess = v.actualConsumed - allowed;
    if (excess <= 0)
      return ok("low", `উৎপাদন ${q(v.finishedProduced)} × coefficient ${q(v.coeffPerUnit)} = অনুমোদিত ব্যবহার ${q(allowed)} একক; প্রকৃত ব্যবহার ${q(v.actualConsumed)} একক — coefficient-এর মধ্যে।`);
    const rev = excess * duty(v, "dutyPerRawUnit");
    return flag("high", excess, "একক", rev,
      `উৎপাদন ${q(v.finishedProduced)} একক × অনুমোদিত coefficient ${q(v.coeffPerUnit)} = অনুমোদিত কাঁচামাল ${q(allowed)} একক; প্রকৃত ব্যবহার ${q(v.actualConsumed)} একক — অতিরিক্ত ${q(excess)} একক। সম্ভাব্য রাজস্ব প্রভাব BDT ${bdt(rev)}।`,
      [{ label: "অনুমোদিত ব্যবহার", value: q(allowed) }, { label: "প্রকৃত ব্যবহার", value: q(v.actualConsumed) }, { label: "অতিরিক্ত", value: q(excess) }]);
  },

  // ২খ) UD-তে দাবিকৃত ব্যবহার vs প্রকৃত রপ্তানি দিয়ে সমর্থিত ব্যবহার
  "num-ud-export"(v) {
    const unsupported = v.udClaimedRaw - v.exportBackedRaw;
    if (unsupported <= 0)
      return ok("low", `UD-তে দাবিকৃত ব্যবহার ${q(v.udClaimedRaw)} একক প্রকৃত রপ্তানি-সমর্থিত ${q(v.exportBackedRaw)} একক দিয়ে সমর্থিত; ব্যত্যয় নেই।`);
    const rev = unsupported * duty(v, "dutyPerRawUnit");
    return flag("high", unsupported, "একক", rev,
      `UD-তে দাবিকৃত কাঁচামাল ব্যবহার ${q(v.udClaimedRaw)} একক, কিন্তু প্রকৃত রপ্তানি (Bill of Export) দিয়ে সমর্থিত মাত্র ${q(v.exportBackedRaw)} একক — অসমর্থিত ${q(unsupported)} একক (রপ্তানি ছাড়াই শুল্কমুক্ত কাঁচামাল ব্যবহার)। সম্ভাব্য রাজস্ব প্রভাব BDT ${bdt(rev)}।`,
      [{ label: "UD দাবি", value: q(v.udClaimedRaw) }, { label: "রপ্তানি-সমর্থিত", value: q(v.exportBackedRaw) }, { label: "অসমর্থিত", value: q(unsupported) }]);
  },

  // ৩) কাঁচামাল material balance — অহিসাবকৃত (সম্ভাব্য স্থানীয় অপসারণ)
  "num-material-balance"(v) {
    const wastage = Number.isFinite(v.wastageAllowedQty) ? v.wastageAllowedQty : 0;
    const available = v.openingStock + v.imported;
    const accounted = v.consumedForExport + wastage + v.closingStock;
    const unaccounted = available - accounted;
    const bd = [
      { label: "প্রাপ্যতা (Opening+Import)", value: q(available) },
      { label: "হিসাবভুক্ত (Export ব্যবহার+অপচয়+Closing)", value: q(accounted) },
      { label: "পার্থক্য", value: q(unaccounted) },
    ];
    if (Math.abs(unaccounted) < 1e-6)
      return ok("low", `কাঁচামাল ব্যালেন্স সঠিক: প্রাপ্যতা ${q(available)} = হিসাবভুক্ত ${q(accounted)} একক; ঘাটতি/উদ্বৃত্ত নেই।`, { breakdown: bd });
    if (unaccounted > 0) {
      const rev = unaccounted * duty(v, "dutyPerRawUnit");
      return flag("high", unaccounted, "একক", rev,
        `Opening ${q(v.openingStock)} + Import ${q(v.imported)} = ${q(available)} একক; হিসাবভুক্ত (রপ্তানি-ব্যবহার ${q(v.consumedForExport)} + অপচয় ${q(wastage)} + Closing ${q(v.closingStock)}) = ${q(accounted)} একক; অহিসাবকৃত ঘাটতি ${q(unaccounted)} একক — সম্ভাব্য শুল্কমুক্ত কাঁচামালের স্থানীয় অপসারণ/বিক্রয়। সম্ভাব্য রাজস্ব প্রভাব BDT ${bdt(rev)}।`,
        bd);
    }
    return { ...flag("medium", -unaccounted, "একক", 0,
      `হিসাবভুক্ত পরিমাণ প্রাপ্যতার চেয়ে ${q(-unaccounted)} একক বেশি — রেকর্ডে অসঙ্গতি (over-accounting)। রেজিস্টার যাচাই করুন; স্বয়ংক্রিয় রাজস্ব ধরা হয়নি।`,
      bd) };
  },

  // ৪) দাবিকৃত অপচয় vs অনুমোদিত অপচয় হার
  "num-wastage"(v) {
    const allowedQty = v.consumedRaw * (v.allowedWastagePct / 100);
    const excess = v.declaredWastageQty - allowedQty;
    if (excess <= 0)
      return ok("low", `অনুমোদিত অপচয় ${q(v.allowedWastagePct)}% × ব্যবহৃত ${q(v.consumedRaw)} = ${q(allowedQty)} একক; দাবিকৃত অপচয় ${q(v.declaredWastageQty)} একক — হারের মধ্যে।`);
    const rev = excess * duty(v, "dutyPerRawUnit");
    return flag("high", excess, "একক", rev,
      `অনুমোদিত অপচয় হার ${q(v.allowedWastagePct)}% × ব্যবহৃত কাঁচামাল ${q(v.consumedRaw)} = অনুমোদিত অপচয় ${q(allowedQty)} একক; দাবিকৃত অপচয় ${q(v.declaredWastageQty)} একক — অতিরিক্ত ${q(excess)} একক (over-wastage)। সম্ভাব্য রাজস্ব প্রভাব BDT ${bdt(rev)}।`,
      [{ label: "অনুমোদিত অপচয়", value: q(allowedQty) }, { label: "দাবিকৃত অপচয়", value: q(v.declaredWastageQty) }, { label: "অতিরিক্ত", value: q(excess) }]);
  },

  // ৫) মেয়াদোত্তীর্ণ (২ বছরের বেশি) বন্ডে থাকা কাঁচামাল — শুল্ক পরিশোধযোগ্য
  "num-overstay"(v) {
    if (v.overstayQty <= 0)
      return ok("low", "মেয়াদোত্তীর্ণ (২ বছরের বেশি) বন্ডে থাকা কাঁচামাল নেই।");
    const rev = v.overstayQty * duty(v, "dutyPerRawUnit");
    return flag("medium", v.overstayQty, "একক", rev,
      `নির্ধারিত মেয়াদ (সাধারণত ২ বছর) অতিক্রান্ত বন্ডে থাকা কাঁচামাল ${q(v.overstayQty)} একক — শুল্ক-কর পরিশোধযোগ্য হয়ে পড়ে। সম্ভাব্য রাজস্ব প্রভাব BDT ${bdt(rev)}।`,
      [{ label: "মেয়াদোত্তীর্ণ পরিমাণ", value: q(v.overstayQty) }]);
  },
};

/** একটি check-এর জন্য ইনপুট map → parsed Number map (খালি → NaN) */
function parseValues(spec, raw) {
  const out = {};
  for (const inp of spec.inputs) {
    const val = raw?.[inp.key];
    out[inp.key] = val === "" || val === null || val === undefined ? NaN : Number(val);
  }
  return out;
}

/**
 * মডিউলের numericChecks স্পেক + auditor-এর ইনপুট থেকে সব check চালায়।
 * @returns {{ results: Array, summary: { flagged:number, totalRevenue:number } }}
 */
export function runNumericChecks(module, inputsByCheck = {}) {
  const specs = module?.numericChecks ?? [];
  const results = specs.map((spec) => {
    const base = { id: spec.id, title: spec.title, area: spec.area, legalRef: spec.legalRef ?? null };
    const compute = COMPUTERS[spec.id];
    if (!compute) return { ...base, status: "insufficient", severity: "low", observation: "এই check-এর হিসাব এখনো যুক্ত হয়নি।", revenueImplication: 0 };
    const values = parseValues(spec, inputsByCheck[spec.id]);
    const g = guard(values, spec);
    if (g) return { ...base, severity: "low", revenueImplication: 0, observation: g.reason, ...g };
    return { ...base, ...compute(values) };
  });
  const flagged = results.filter((r) => r.status === "flag");
  const totalRevenue = flagged.reduce((s, r) => s + Number(r.revenueImplication || 0), 0);
  return { results, summary: { flagged: flagged.length, totalRevenue: round2(totalRevenue) } };
}

/** flag হওয়া একটি numeric result → finding payload (storage.addFinding-এর শেপে) */
export function resultToFinding(result) {
  return {
    checklistId: null,
    area: result.area,
    title: result.title,
    observation: result.observation,
    legalRef: result.legalRef ?? null,
    revenueImplication: Number(result.revenueImplication || 0),
    severity: result.severity ?? "medium",
    source: "numeric",
  };
}

File: customs-bond-vat-copilot/backend/services/checks/numeric.py
"""
সংখ্যাগত auto-check ইঞ্জিন — বন্ড অডিটের reconciliation (JS src/checks/numeric.js হইতে পোর্ট)।

নীতি (নিরীক্ষক-নির্ধারিত):
  • প্রতিটি চেক দ্বিমুখী — ঘাটতি (রাজস্ব দাবি) ও over-recording (রেকর্ড অসঙ্গতি, রাজস্ব ০) আলাদা।
  • শুল্ক-ইনপুট ঐচ্ছিক — না দিলে শুধু পরিমাণ-ব্যত্যয়, রাজস্ব ০।
  • ★ দ্বৈত গণনা রোধ: এই উপমোট নিজে থেকে চূড়ান্ত রাজস্ব-মোটে যোগ হয় না; শুধু
    নিরীক্ষক-গৃহীত Finding যোগ হয় (result_to_finding → addFinding)।

প্রতিটি compute একটি values dict নেয় (missing → None), result দেয়:
  { status, severity, discrepancy, unit, revenue_implication, observation, breakdown }
  status: "ok" | "flag" | "insufficient"
"""
from __future__ import annotations

from typing import Any, Callable, Optional


def _r2(n: float) -> float:
    return round(float(n) + 0.0, 2)


def _isnum(v: Any) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _bdt(n: float) -> str:
    return f"{float(n or 0):,.0f}"


def _q(n: float) -> str:
    return f"{_r2(n):,.2f}"


def _duty(v: dict, key: str) -> float:
    """duty ইনপুট ঐচ্ছিক; না দিলে ০ ধরে শুধু পরিমাণ-ব্যত্যয় দেখায়।"""
    return float(v[key]) if _isnum(v.get(key)) else 0.0


def _ok(severity: str, observation: str, breakdown=None) -> dict:
    return {"status": "ok", "severity": severity or "low", "observation": observation,
            "revenue_implication": 0.0, "breakdown": breakdown}


def _flag(severity: str, discrepancy: float, unit: str, revenue: float,
          observation: str, breakdown=None) -> dict:
    return {"status": "flag", "severity": severity, "discrepancy": _r2(discrepancy),
            "unit": unit, "revenue_implication": _r2(revenue),
            "observation": observation, "breakdown": breakdown}


# ---------------- per-check computers ----------------

def _c_entitlement(v: dict) -> dict:
    excess = v["imported"] - v["approvedEntitlement"]
    if excess <= 0:
        return _ok("low", f"আমদানি {_q(v['imported'])} একক অনুমোদিত entitlement "
                          f"{_q(v['approvedEntitlement'])} একক-এর মধ্যে; breach নেই।")
    rev = excess * _duty(v, "dutyPerUnit")
    return _flag("high", excess, "একক", rev,
                 f"অনুমোদিত entitlement/UP {_q(v['approvedEntitlement'])} একক-এর বিপরীতে "
                 f"আমদানি {_q(v['imported'])} একক — অতিরিক্ত {_q(excess)} একক (entitlement breach)। "
                 f"সম্ভাব্য রাজস্ব প্রভাব BDT {_bdt(rev)}।",
                 [{"label": "অনুমোদিত entitlement", "value": _q(v['approvedEntitlement'])},
                  {"label": "আমদানি", "value": _q(v['imported'])},
                  {"label": "অতিরিক্ত", "value": _q(excess)}])


def _c_coefficient(v: dict) -> dict:
    allowed = v["finishedProduced"] * v["coeffPerUnit"]
    excess = v["actualConsumed"] - allowed
    if excess <= 0:
        return _ok("low", f"উৎপাদন {_q(v['finishedProduced'])} × coefficient {_q(v['coeffPerUnit'])} "
                          f"= অনুমোদিত ব্যবহার {_q(allowed)} একক; প্রকৃত {_q(v['actualConsumed'])} একক — সীমার মধ্যে।")
    rev = excess * _duty(v, "dutyPerRawUnit")
    return _flag("high", excess, "একক", rev,
                 f"উৎপাদন {_q(v['finishedProduced'])} × অনুমোদিত coefficient {_q(v['coeffPerUnit'])} "
                 f"= অনুমোদিত কাঁচামাল {_q(allowed)} একক; প্রকৃত ব্যবহার {_q(v['actualConsumed'])} একক — "
                 f"অতিরিক্ত {_q(excess)} একক। সম্ভাব্য রাজস্ব প্রভাব BDT {_bdt(rev)}।",
                 [{"label": "অনুমোদিত ব্যবহার", "value": _q(allowed)},
                  {"label": "প্রকৃত ব্যবহার", "value": _q(v['actualConsumed'])},
                  {"label": "অতিরিক্ত", "value": _q(excess)}])


def _c_ud_export(v: dict) -> dict:
    unsupported = v["udClaimedRaw"] - v["exportBackedRaw"]
    if unsupported <= 0:
        return _ok("low", f"UD/EP-তে দাবিকৃত ব্যবহার {_q(v['udClaimedRaw'])} একক প্রকৃত রপ্তানি-সমর্থিত "
                          f"{_q(v['exportBackedRaw'])} একক দিয়ে সমর্থিত; ব্যত্যয় নেই।")
    rev = unsupported * _duty(v, "dutyPerRawUnit")
    return _flag("high", unsupported, "একক", rev,
                 f"UD/EP (ইপিজেড হলে EP/Sales Contract)-এ দাবিকৃত ব্যবহার {_q(v['udClaimedRaw'])} একক, "
                 f"কিন্তু রপ্তানি (Bill of Export) দিয়ে সমর্থিত মাত্র {_q(v['exportBackedRaw'])} একক — "
                 f"অসমর্থিত {_q(unsupported)} একক (রপ্তানি ছাড়াই শুল্কমুক্ত ব্যবহার)। "
                 f"সম্ভাব্য রাজস্ব প্রভাব BDT {_bdt(rev)}।",
                 [{"label": "UD/EP দাবি", "value": _q(v['udClaimedRaw'])},
                  {"label": "রপ্তানি-সমর্থিত", "value": _q(v['exportBackedRaw'])},
                  {"label": "অসমর্থিত", "value": _q(unsupported)}])


def _c_material_balance(v: dict) -> dict:
    wastage = float(v["wastageAllowedQty"]) if _isnum(v.get("wastageAllowedQty")) else 0.0
    available = v["openingStock"] + v["imported"]
    accounted = v["consumedForExport"] + wastage + v["closingStock"]
    unaccounted = available - accounted
    bd = [{"label": "প্রাপ্যতা (Opening+Import)", "value": _q(available)},
          {"label": "হিসাবভুক্ত (Export+অপচয়+Closing)", "value": _q(accounted)},
          {"label": "পার্থক্য", "value": _q(unaccounted)}]
    if abs(unaccounted) < 1e-6:
        return _ok("low", f"কাঁচামাল ব্যালেন্স সঠিক: প্রাপ্যতা {_q(available)} = হিসাবভুক্ত "
                          f"{_q(accounted)} একক; ঘাটতি/উদ্বৃত্ত নেই।", bd)
    if unaccounted > 0:
        rev = unaccounted * _duty(v, "dutyPerRawUnit")
        return _flag("high", unaccounted, "একক", rev,
                     f"Opening {_q(v['openingStock'])} + Import {_q(v['imported'])} = {_q(available)} একক; "
                     f"হিসাবভুক্ত (রপ্তানি-ব্যবহার {_q(v['consumedForExport'])} + অপচয় {_q(wastage)} + "
                     f"Closing {_q(v['closingStock'])}) = {_q(accounted)} একক; অহিসাবকৃত ঘাটতি "
                     f"{_q(unaccounted)} একক — সম্ভাব্য শুল্কমুক্ত কাঁচামালের স্থানীয় অপসারণ/বিক্রয়। "
                     f"সম্ভাব্য রাজস্ব প্রভাব BDT {_bdt(rev)}।", bd)
    return _flag("medium", -unaccounted, "একক", 0.0,
                 f"হিসাবভুক্ত পরিমাণ প্রাপ্যতার চেয়ে {_q(-unaccounted)} একক বেশি — রেকর্ডে অসঙ্গতি "
                 f"(over-accounting)। রেজিস্টার যাচাই করুন; স্বয়ংক্রিয় রাজস্ব ধরা হয়নি।", bd)


def _c_wastage(v: dict) -> dict:
    allowed_qty = v["consumedRaw"] * (v["allowedWastagePct"] / 100)
    excess = v["declaredWastageQty"] - allowed_qty
    if excess <= 0:
        return _ok("low", f"অনুমোদিত অপচয় {_q(v['allowedWastagePct'])}% × ব্যবহৃত {_q(v['consumedRaw'])} "
                          f"= {_q(allowed_qty)} একক; দাবিকৃত {_q(v['declaredWastageQty'])} একক — হারের মধ্যে।")
    rev = excess * _duty(v, "dutyPerRawUnit")
    return _flag("high", excess, "একক", rev,
                 f"অনুমোদিত অপচয় হার {_q(v['allowedWastagePct'])}% × ব্যবহৃত {_q(v['consumedRaw'])} "
                 f"= অনুমোদিত অপচয় {_q(allowed_qty)} একক; দাবিকৃত {_q(v['declaredWastageQty'])} একক — "
                 f"অতিরিক্ত {_q(excess)} একক (over-wastage)। সম্ভাব্য রাজস্ব প্রভাব BDT {_bdt(rev)}।",
                 [{"label": "অনুমোদিত অপচয়", "value": _q(allowed_qty)},
                  {"label": "দাবিকৃত অপচয়", "value": _q(v['declaredWastageQty'])},
                  {"label": "অতিরিক্ত", "value": _q(excess)}])


def _c_overstay(v: dict) -> dict:
    if v["overstayQty"] <= 0:
        return _ok("low", "মেয়াদোত্তীর্ণ (২ বছরের বেশি) বন্ডে থাকা কাঁচামাল নেই।")
    rev = v["overstayQty"] * _duty(v, "dutyPerRawUnit")
    return _flag("medium", v["overstayQty"], "একক", rev,
                 f"নির্ধারিত মেয়াদ (working default ২ বছর — gazette citation অপেক্ষমাণ) অতিক্রান্ত বন্ডে "
                 f"থাকা কাঁচামাল {_q(v['overstayQty'])} একক — শুল্ক-কর পরিশোধযোগ্য। "
                 f"সম্ভাব্য রাজস্ব প্রভাব BDT {_bdt(rev)}।",
                 [{"label": "মেয়াদোত্তীর্ণ পরিমাণ", "value": _q(v['overstayQty'])}])


def _be_vs_register(label: str) -> Callable[[dict], dict]:
    def _compute(v: dict) -> dict:
        diff = v["beQty"] - v["registerQty"]
        bd = [{"label": f"{label} B/E আমদানি", "value": _q(v['beQty'])},
              {"label": "রেজিস্টারে লিপিবদ্ধ", "value": _q(v['registerQty'])},
              {"label": "পার্থক্য", "value": _q(diff)}]
        if abs(diff) < 1e-6:
            return _ok("low", f"{label}: B/E {_q(v['beQty'])} একক = রেজিস্টারে লিপিবদ্ধ "
                              f"{_q(v['registerQty'])} একক; মিল আছে।", bd)
        if diff > 0:
            rev = diff * _duty(v, "dutyPerUnit")
            return _flag("high", diff, "একক", rev,
                         f"{label} আমদানি: Bill of Entry অনুযায়ী {_q(v['beQty'])} একক, কিন্তু রেজিস্টারে "
                         f"লিপিবদ্ধ {_q(v['registerQty'])} একক — {_q(diff)} একক রেজিস্টারভুক্ত হয়নি "
                         f"(unrecorded; সম্ভাব্য অপসারণ)। সম্ভাব্য রাজস্ব প্রভাব BDT {_bdt(rev)}।", bd)
        return _flag("medium", -diff, "একক", 0.0,
                     f"{label} আমদানি: রেজিস্টারে লিপিবদ্ধ {_q(v['registerQty'])} একক B/E {_q(v['beQty'])} "
                     f"একক-এর চেয়ে {_q(-diff)} একক বেশি — over-recording/রেকর্ড অসঙ্গতি; যাচাই করুন। "
                     f"স্বয়ংক্রিয় রাজস্ব ধরা হয়নি।", bd)
    return _compute


COMPUTERS: dict[str, Callable[[dict], dict]] = {
    "num-entitlement": _c_entitlement,
    "num-coefficient": _c_coefficient,
    "num-ud-export": _c_ud_export,
    "num-material-balance": _c_material_balance,
    "num-wastage": _c_wastage,
    "num-overstay": _c_overstay,
    "num-be-register-raw": _be_vs_register("কাঁচামাল"),
    "num-be-register-machinery": _be_vs_register("মেশিনারিজ"),
    "num-be-register-sample": _be_vs_register("Sample"),
}


# ---------------- runner ----------------

def _parse_values(spec: dict, raw: Optional[dict]) -> dict:
    raw = raw or {}
    out: dict[str, Any] = {}
    for inp in spec["inputs"]:
        val = raw.get(inp["key"])
        if val in ("", None):
            out[inp["key"]] = None
        else:
            try:
                out[inp["key"]] = float(val)
            except (TypeError, ValueError):
                out[inp["key"]] = None
    return out


def _guard(values: dict, spec: dict) -> Optional[dict]:
    keys = [i["key"] for i in spec["inputs"]]
    if not any(_isnum(values.get(k)) for k in keys):
        return {"status": "insufficient", "reason": "কোনো সংখ্যা দেওয়া হয়নি"}
    missing = [i["label"] for i in spec["inputs"]
               if not i.get("optional") and not _isnum(values.get(i["key"]))]
    if missing:
        return {"status": "insufficient", "reason": "প্রয়োজনীয় তথ্য অনুপস্থিত: " + ", ".join(missing)}
    return None


def run_numeric_checks(specs: list[dict], inputs_by_check: Optional[dict] = None) -> dict:
    """specs + auditor ইনপুট → সব check চালায়।"""
    inputs_by_check = inputs_by_check or {}
    results = []
    for spec in specs:
        base = {"id": spec["id"], "title": spec.get("title", ""), "area": spec.get("area", ""),
                "legalRef": spec.get("legalRef")}
        compute = COMPUTERS.get(spec["id"])
        if not compute:
            results.append({**base, "status": "insufficient", "severity": "low",
                            "observation": "এই check-এর হিসাব এখনো যুক্ত হয়নি।", "revenue_implication": 0.0})
            continue
        values = _parse_values(spec, inputs_by_check.get(spec["id"]))
        g = _guard(values, spec)
        if g:
            results.append({**base, "severity": "low", "revenue_implication": 0.0,
                            "observation": g["reason"], "status": "insufficient"})
            continue
        results.append({**base, **compute(values)})
    flagged = [r for r in results if r["status"] == "flag"]
    total = _r2(sum(float(r.get("revenue_implication") or 0) for r in flagged))
    return {"results": results, "summary": {"flagged": len(flagged), "totalRevenue": total}}


def result_to_finding(result: dict) -> dict:
    """flag হওয়া numeric result → finding payload (source=numeric)।"""
    return {
        "checklistId": None,
        "area": result.get("area", ""),
        "title": result.get("title", ""),
        "observation": result.get("observation", ""),
        "legalRef": result.get("legalRef"),
        "revenueImplication": float(result.get("revenue_implication") or 0),
        "severity": result.get("severity", "medium"),
        "source": "numeric",
    }


__all__ = ["run_numeric_checks", "result_to_finding", "COMPUTERS"]
